### PR TITLE
Api interceptor implementation to allow app developers to peek responses

### PR DESCRIPTION
**Pull request by onurays**

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import okhttp3.Interceptor
+import okhttp3.Response
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Interceptor class for provided api paths.
+ */
+@Singleton
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt:28]: In SDK we use @MatrixScope instead

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiPath.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.api.network
+
+import org.matrix.android.sdk.internal.network.NetworkConstants
+
+enum class ApiPath(val path: String, val method: String) {
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiPath.kt:21]: I think it's error prone as we have to add path in different places each time we add a new API.
Maybe we could generate the paths from the Retrofit interfaces.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/di/NetworkModule.kt [hunk 3]
```diff
@@ -54,6 +55,13 @@ internal object NetworkModule {
         return CurlLoggingInterceptor()
     }
 
+    @MatrixScope
+    @Provides
+    @JvmStatic
+    fun providesApiInterceptor(): ApiInterceptor {
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/di/NetworkModule.kt:61]: This is not necessary to add provides method when you already have Inject annotation on the constructor of the classe. So you can remove those lines ^

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt (hunk 1 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiPath.kt (hunk 2 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiPath.kt:21]: I agree, we could improve that in the future

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        val request = chain.request()
+        val path = request.url.encodedPath.replaceFirst("/", "")
+        val method = request.method
+
+        val response = chain.proceed(request)
+
+        findApiPath(path, method)?.let { apiPath ->
+            response.peekBody(Long.MAX_VALUE).string().let {
+                apiResponseListenersMap[apiPath]?.forEach { listener ->
+                    listener.onApiResponse(apiPath, it)
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    interface Listener {
+        fun onApiResponse(path: ApiPath, response: String)
+    }
+
+    init {
+        Timber.d("ApiInterceptor.init")
+    }
+
+    private val apiResponseListenersMap = mutableMapOf<ApiPath, MutableList<Listener>>()
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/network/ApiInterceptor.kt]: Done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/di/NetworkModule.kt (hunk 3 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/di/NetworkModule.kt]: removed

**Review by ganfra — APPROVED** [review]
Thanks, for the changes, LGTM now

**Pull request merged**